cocos/renderer/backend/opengl/ProgramGL.cpp

```diff
@@ -37,16 +37,16 @@ CC_BACKEND_BEGIN
 ProgramGL::ProgramGL(const std::string& vertexShader, const std::string& fragmentShader)
 : Program(vertexShader, fragmentShader)
 {
-    _vertexShaderModule = static_cast<ShaderModuleGL*>(ShaderCache::newVertexShaderModule(_vertexShader));
 #if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
-    std::string headersDef("#ifdef GL_FRAGMENT_PRECISION_HIGH\n" \
-                           "precision highp float;\n precision highp int;\n" \
-                           "#else\n" \
-                           "precision mediump float;\n precision mediump int;\n"
-                           "#endif\n");
-    headersDef.append(fragmentShader);
-    _fragmentShaderModule = static_cast<ShaderModuleGL*>(ShaderCache::newFragmentShaderModule(headersDef));
+    //some device required manually specify the precision qualifiers for vertex shader.
+    std::string vsPreDefine("#version 100\n precision highp float;\n precision highp int;\n");
+    std::string fsPreDefine("precision mediump float;\n precision mediump int;\n");
```

Comment:
vsPreDefine and fsPreDefine are const strings, it is better to make them as const member variables.

Reply:
I think `vsPreDefine` is better as a non-const string since it will append with shader source.

Reply:
But you can insert vsPreDefine at the begin of the shader.

Reply:
Ok, it is not suitable, ignore it.